### PR TITLE
chore: sharding tests update

### PR DESCRIPTION
**Pull request by romanzac**
### Description
Batch of tests for Waku sharding functionality to extend existing tests 

### Changes
Tests available to run:
cd go-waku/waku/v2/node
go test

### CODE REVIEW

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:348]: If we are writing static sharding tests, we should specify clusterID when we create the Waku node.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:355]: ClusterID 0 cannot be used for static sharding. 0 would be considered as non-sharded default pubsubTopic.

Refer rfc https://rfc.vac.dev/spec/52/ for more details on shard allocation and usage.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:407]: This test looks like testing Waku relay with static sharding and it looks like only 1 node is used for the same. 
Not sure what additional code coverage are we achieving with this.

Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:407]: I am learning and creating rather than trying to improve coverage in this PR. I saw some tests for sharding already and would not like to do the same things. I've marked the PR ready to review because I would need your help. We would eventually move out some of the tests once we vet them. Please note I mostly use public API. Even, if it was not indented this way,  it can still yield insights like https://github.com/waku-org/go-waku/issues/1070

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:355]: Are all 0-15 cluster index IDs reserved -> cannot be used with static sharding ?

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:348]: I see, so the clusterID could actually live, be useful, without sharded topic ?

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:355]: Reserved means can be used with static-sharding. It is just that someone else should not use these clusters other than Status/Waku.
But 0 is not supposed to be used for static-sharding as it is marked as global.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:407]: If it is still in progress, then you can leave the PR in draft and assign me for review and ping me so that i can take a look.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:348]: Did not get what you mean

Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:348]: I curious about value of setting clusterID (node.WithClusterID ) without having intent to use any shard.

Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:348]: But you are using a shard when you are subscribing the pubsubTopic that is part of the clusterid. 
the pubsubTopic format is `/waku/2/rs/<cluster-ID>/<shard>` which means a shard will be included.

Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:348]: I see, it is used to create metadata when node instance is created: 

metadata := metadata.NewWakuMetadata(w.opts.clusterID, w.localNode, w.log)
w.metadata = metadata

Metadata won't even start without clusterID != 0:
```
// Start inits the metadata protocol
func (wakuM *WakuMetadata) Start(ctx context.Context) error {
	if wakuM.clusterID == 0 {
		wakuM.log.Warn("no clusterID is specified. Protocol will not be initialized")
		return nil
	}
	.
	.
	.
}
```
I've added WithClusterID and it has no effect on my test. Next interesting thing: node is a member of only one cluster from metadata perspective. Correct ? https://github.com/waku-org/go-waku/pull/1060/commits/4266513a252b3b30480bf2ccad6a85e57d7ee7f0

Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:348]: `node is a member of only one cluster from metadata perspective. Correct ?`

Yes, a node can be part of a single cluster. And without clusterID metadata protocol won't be initialized.

As i mentioned above, since these are tests for sharding and as Waku supports 2 types static and auto sharding, we should try to use those formats.

**Comment by romanzac** [comment]
Issue https://github.com/waku-org/go-waku/issues/1077 related to this PR

File: tests/utils.go
Comment (chaitanyaprem) [inline tests/utils.go:415]: minor suggest, maybe take timeout as an arg to the method so that it can be used in other tests as well.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:342]: Shouldn't 20 be passed as clusterID as node1 is being initialised with clusterID as 20?

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:346]: Same as above

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:398]: Using clusterID 0 here? 
Better to use a different pubsunTopic in cluster 20 itself which is not subscribed.

**Review by chaitanyaprem — COMMENTED** [review]
Did partial review, but looks like comments need to be addressed

File: tests/utils.go
Comment (romanzac) [reply tests/utils.go]: Thanks for suggestion. Added timeout param to both WaitFor funcs. https://github.com/waku-org/go-waku/pull/1060/commits/99e3f334cd8356d343d3f3f7ac750ca46508de05

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go]: It is on purpose to check if sharded cluster ID could be different than native clusterID - message still delivered. I've added variables to display intent. Not sure if "native" and sharded are the good labels. https://github.com/waku-org/go-waku/pull/1060/commits/99e3f334cd8356d343d3f3f7ac750ca46508de05

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go]: Good point, I've forgot 0 is already subscribed. Fixed at https://github.com/waku-org/go-waku/pull/1060/commits/99e3f334cd8356d343d3f3f7ac750ca46508de05

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go]: Hmm, that is because clusterID is just a user abstraction for underlying pubsubTopic. 
When we do subscribe for relay, we don't check if the node is initialized with that clusterID.
Not sure if we will add such a check in go-waku, but it would be nice to not allow this.

cc @richard-ramos, something to be considered for nwaku bindings.

Comment (romanzac) [reply waku/v2/node/wakunode2_test.go]: I wonder what is the story behind the lonely clusterID ?  Topic is the central concept we use to shard the entire network isn't it ? Why do we need on the side clusterID which is just one of many possible topics ?

**Comment by romanzac** [comment]
I've reverted "TestStaticShardingLimits" to its desired form and marked it as flaky. This way we can unblock this PR and start a collection of flaky tests. Flaky test could take a time to resolve, sometimes it could lead to a design change. There is still some value to keep them. What are your thoughts ? @chaitanyaprem @richard-ramos 
I suggest skip mechanism for those tests would be:
```
	if os.Getenv("RUN_FLAKY_TESTS") != "true" {

		log.Info("Skipping", zap.String("test", t.Name()),
			zap.String("reason", "RUN_FLAKY_TESTS environment variable is not set to true"))
		t.SkipNow()
	}
```

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:424]: Looks like a nice test, but maybe we may want to add more verification by sending messages in a loop randomly on a topic and waiting for them to be received to ensure all subscriptions are valid.

File: waku/v2/protocol/filter/filter_test.go
Comment (chaitanyaprem) [inline waku/v2/protocol/filter/filter_test.go:562]: Again, as i had mentioned above...whatever we want can be achieved by using shards from same cluster. 

Let us not use shards from different cluster.

File: waku/v2/protocol/lightpush/waku_lightpush_test.go
Comment (chaitanyaprem) [inline waku/v2/protocol/lightpush/waku_lightpush_test.go:330]: Node2 doesn't need to be a relay node if it wants to use lightpush for publishing. 
It can just be a simple lightpush node and use node1 to publish.

File: waku/v2/protocol/store/waku_store_protocol_test.go
Comment (chaitanyaprem) [inline waku/v2/protocol/store/waku_store_protocol_test.go:473]: Did not understand why this is required? DefaultpubsubTopic is not related to static sharding.

**Review by chaitanyaprem — CHANGES_REQUESTED** [review]
Gave some comments.

Also CI is failing which needs to be fixed.

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:331]: I picked 2 as a minimum for the assumed use case: 
When an operator starts node, they have a control about cluster ID. They don't have any control what pubsub topic with other cluster IDs their users might subscribe to. Act of subscription is anonymous and is not (spam) protected. Provided user follows the supported pubsub topic spec.

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:359]: Described above. If this usage is not intended by devs, it would be better to prevent it in the code. 

Is the cluster ID an administrative concept ? More like a sticky label of a node for human operator?  ENR is hard to remember, otherwise ENR+pubsub topic does the real work with message routing. Am I getting it?

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:411]: Here I am checking, if filtering based on content topic works within the context of static sharding.

File: waku/v2/protocol/filter/filter_test.go
Comment (romanzac) [reply waku/v2/protocol/filter/filter_test.go]: Cluster IDs unified at https://github.com/waku-org/go-waku/pull/1060/commits/7404bb416fe513d2dd68387657767adaa4f5b852

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go]: That is to help with discovery of peers. If we want to scale beyond 1 pubsubTopic, we can do that by having multiple pubsubTopics. But, how do we discover peers in the network then?
We use discv5 for discovery which depends on ENR. ENR has a size restriction where-in we can't include a list of pubsubTopics in it. 

Hence to make it easier to encode this information in the ENR, the pubsubTopic has been formatted into clusterID and shards. 
The idea of clusterID is kind of similar to networkID in a blockchain , where-in a node is expected to be part of 1 cluster only.
You can refer to rfc https://github.com/waku-org/specs/blob/master/standards/core/relay-sharding.md for more details. 

The limit of 1024 shards within a cluster also has similar reasons behind it.  One of them being discv5 DHT not being able to scale beyond a particular number of nodes i.e finding nodes beyond a particular size is going to take a lot of time that it becomes almost useless. IIRC that node count is 10K or something like that. 


Hope this helps.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:331]: Ideally, we don't want users doing that. i.e start node with clusterID x and subscribe to a topic that is belonging to clusterID y.  
Wondering if we should throw an error if someone tries to subscribe to a topic from different cluster.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:359]: Gave a brief explanation https://github.com/waku-org/go-waku/pull/1060#discussion_r1587048424

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:411]: Ok, in that case we should subscribe to a specific contentTopic  ct1 within a pubsubTopic and then send messages to ct1 and some messages to ct2. 
Add a check that msgs sent to ct1 were received by intended node and ct2 messages were not received (i.e filtered out).

File: waku/v2/protocol/store/waku_store_protocol_test.go
Comment (romanzac) [reply waku/v2/protocol/store/waku_store_protocol_test.go:473]: Is there another way how can I get empty subscription ? I need it for s2.Start(ctx, sub1), but I wish s2 doesn't know anything upfront about pubsub topics of s1.

Comment (chaitanyaprem) [reply waku/v2/protocol/store/waku_store_protocol_test.go:473]: Not sure what you mean by empty sub, but you can always ise a different pubsubTopic in same cluster that is not being used by s1.

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go]: Thanks for explanation.

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:331]: I would throw an error with a cool message and reject to subscribe. Misuse, abuse or any misunderstanding like mine would create future work, which scales too well. I think it is sensible to have a lean design at this stage of Waku's  adoption.

File: waku/v2/protocol/store/waku_store_protocol_test.go
Comment (romanzac) [reply waku/v2/protocol/store/waku_store_protocol_test.go]: New topic added at https://github.com/waku-org/go-waku/pull/1060/commits/12a8923d6270d4e185a683607bb51fe1d3beb987

File: waku/v2/protocol/lightpush/waku_lightpush_test.go
Comment (romanzac) [reply waku/v2/protocol/lightpush/waku_lightpush_test.go]: Node2 is a forwarder to node1, which is only relay.  I've tried to change node2 to be only lightpush, and received confusing error. https://github.com/waku-org/go-waku/pull/1060/commits/73a90e085c8b90418abab36e2abea17abee74d58

```
--- FAIL: TestWakuLightPushWithStaticSharding (2.47s)
    waku_lightpush_test.go:368:
        	Error Trace:	/Users/roman/sources/waku-org/go-waku/waku/v2/protocol/lightpush/waku_lightpush_test.go:368
        	Error:      	Received unexpected error:
        	            	failed to negotiate protocol: protocols not supported: [/vac/waku/lightpush/2.0.0-beta1]
        	Test:       	TestWakuLightPushWithStaticSharding
```

Have I missed something ?

Comment (chaitanyaprem) [reply waku/v2/protocol/lightpush/waku_lightpush_test.go]: Ah..ok. 
If so, then node2 has to be support relay and lightpush as service node.

Comment (romanzac) [reply waku/v2/protocol/lightpush/waku_lightpush_test.go:337]: Reorganized at https://github.com/waku-org/go-waku/pull/1060/commits/093ea363a6ac347c0c424360a34075bfed62a4a7

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:424]: Would not be sufficient if we randomly chose one pubsub topic at each test run ?

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:411]: I have three pubsub and two content topics in the test. ct1 and ct2 are subscribed together with their respective pt1 and pt2. Cases: 1. pt1+ct1 subscribed and receiving, 2. pt3(non-subscribed)+ct2 not receiving,  and 3. pt1+ct2(non-subscribed, but known to the node) not receiving. 

I guess first and third case matches with your proposal except I have a twist, the ct2 is not completely unknown to the node ?

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go]: I've unified clusterID with pubsub topic at https://github.com/waku-org/go-waku/pull/1060/commits/dd960ef93dc1c72149e3561d326c77f4721b57b4

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:423]: It should be, but maybe just randomly choose some shards to ensure no weird bug in the code.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:358]: Also, you can raise an issue so that we don't forget about this.

Although not sure if we want to implement in go-waku as it is being sunset.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [inline waku/v2/node/wakunode2_test.go:402]: Better to use pubsubTopic from testCluster 20 here which is not subscribed to.

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:410]: Ok, that is a weird case but fine.

Ideally a contentTopic is tied to a pubsubTopic, not sure there would be an actual usecase where the same contentTopic is used in multiple pubsubTopics

File: waku/v2/node/wakunode2_test.go
Comment (romanzac) [reply waku/v2/node/wakunode2_test.go:358]: Let's keep it in this PR only. I don't want to generate additional work, if that is what sunsetting means.

**Comment by chaitanyaprem** [comment]
@romanzac CI seems to be failing, we will have to fix that. 

One more comment and once both of these are fixed, we can merge .

File: waku/v2/node/wakunode2_test.go
Comment (chaitanyaprem) [reply waku/v2/node/wakunode2_test.go:358]: My bad, no issue with this comment. 

I was talking about https://github.com/waku-org/go-waku/pull/1060#discussion_r1587172611.

**Comment by romanzac** [comment]
> @romanzac CI seems to be failing, we will have to fix that.
> 
> One more comment and once both of these are fixed, we can merge .

CI if failing at Go installation step. I saw https://github.com/waku-org/go-waku/pull/1091 upgrades Go version. Could you please check for me value of ${{ needs.env.outputs.go_version }} in repo settings or give me access so I can see the value ?

**Comment by chaitanyaprem** [comment]
> > @romanzac CI seems to be failing, we will have to fix that.
> > One more comment and once both of these are fixed, we can merge .
> 
> CI if failing at Go installation step. I saw #1091 upgrades Go version. Could you please check for me value of ${{ needs.env.outputs.go_version }} in repo settings or give me access so I can see the value ?

Hmm, that is not yet merged. Also, wondering why it is failing for this PR only?

Also, i don't see any such setting in repo settings->environment. Is that the place to look?

**Review by chaitanyaprem — APPROVED** [review]
Can be merged once CI is fixed.

**Comment by chaitanyaprem** [comment]
Maybe @richard-ramos know whats happening with CI.

**Comment by romanzac** [comment]
> Can be merged once CI is fixed.

I've contacted Richard few hours ago. I'll try few things before he responds.

**Comment by chaitanyaprem** [comment]
Why is this branch showing soo much diff with master? 
Let us not merge when diff is like this..it means the gut tree is probably messed up. 
Ideally, we should rebase from master, pass CI and then merge it. 

Hopefully nothing got messed up.

**Comment by romanzac** [comment]
> Why is this branch showing soo much diff with master? Let us not merge when diff is like this..it means the git tree is probably messed up. Ideally, we should rebase from master, pass CI and then merge it.
> 
> Hopefully nothing got messed up.

This is what I have done. Rebased from master, passed CI and then merged. No worries.

**Comment by chaitanyaprem** [comment]
> > Why is this branch showing soo much diff with master? Let us not merge when diff is like this..it means the git tree is probably messed up. Ideally, we should rebase from master, pass CI and then merge it.
> > 
> > Hopefully nothing got messed up.
> 
> This is what I have done. Rebased from master, passed CI and then merged. No worries. 

Oh, interesting...in that case the diff shouldn't be this big right

**Comment by romanzac** [comment]
> > > Why is this branch showing soo much diff with master? Let us not merge when diff is like this..it means the git tree is probably messed up. Ideally, we should rebase from master, pass CI and then merge it.
> > > Hopefully nothing got messed up.
> > 
> > 
> > This is what I have done. Rebased from master, passed CI and then merged. No worries.
> 
> Oh, interesting...in that case the diff shouldn't be this big right

I remember I've done merge at some point too.

**Pull request merged**
